### PR TITLE
Spawn crates /w Warheads & minor crate improvements

### PR DESCRIPTION
**Pull request by Starkku**
## Spawn powerup crate

- Warheads can now spawn powerup crates of specified type(s) on their impact cells (if free, or nearby cells if occupied something other than a crate) akin to map trigger action 108 ('Create Crate').
  - `SpawnsCrateN` where N is a number starting from 0, parsed until no key is found can be used to define the type of crate spawned.
  - `SpawnsCrateN.Weight` is a number that determines relative weighting of spawning corresponding crate type vs. other listed ones (0 is no chance, higher means higher probability) defaulting to 1 if not defined.
  - `SpawnsCrate.Type/Weight` is an alias for `SpawnsCrate0.Type/Weight` if latter is not set.

In `rulesmd.ini`:
```ini
[SOMEWARHEAD]            ; Warhead
SpawnsCrate(N).Type=     ; Powerup crate type enum (money|unit|healbase|cloak|explosion|napalm|squad|reveal|armor|speed|firepower|icbm|invulnerability|veteran|ionstorm|gas|tiberium|pod)
SpawnsCrate(N).Weight=1  ; integer
```

## Crate improvements

There are some improvements on goodie crate logic:
- The limit of vehicles a player can own before unit crates start giving money instead can now be customized by setting `UnitCrateVehicleCap`. Negative numbers disable the cap entirely.
- `FreeMCV` setting is now actually respected and can be used to disable the forced unit selected from `[General]` -> `BaseUnit` that is given if player picks a crate and has enough credits but no existing buildings or `BaseUnit` vehicles.
  - The previously hardcoded credits threshold that must be passed can also now be customized via `FreeMCV.CreditsThreshold`.
- It is possible to influence weighting of units given from crates (`CrateGoodie=true`) via `CrateGoodie.RerollChance`, which determines the chance that if this type of unit is rolled, it will reroll again for another type of unit.

In `rulesmd.ini`:
```ini
[CrateRules]
UnitCrateVehicleCap=50         ; integer
FreeMCV=true                   ; boolean
FreeMCV.CreditsThreshold=1500  ; integer

[SOMEVEHICLE]                  ; VehicleType
CrateGoodie.RerollChance=0.0   ; floating point value, percents or absolute (0.0-1.0)
```

### CODE REVIEW

**Comment by mbnq** [comment]
Thats very cool!

**Comment by mbnq** [comment]
I love this and thank you for making this!

```
money:		ok
unit:		ok
healbase: 	ok
cloak:		ok
explosion:	ok
napalm:		ok
squad:		money
reveal:		ok
armor:		ok
speed:		ok
firepower:	ok
icbm:		ok
invulnerability: displays armor p-up animation, nothing happens
veteran:	ok
ionstorm:	nothing
gas:		ok
pod:		nothing
darkness:	crate does not spawn
```

Crates spawned on the water are money, as normal. 
Squad does't work and [it shouldn't i think](https://modenc.renegadeprojects.com/Powerups).
Invulnerability displays armor power up anim, [nothing happens.](https://modenc.renegadeprojects.com/Powerups)
Pod no animation, [no effect](https://modenc.renegadeprojects.com/Powerups). 

I would suggest to default crate to moneycrate if crate is not supported, 
missing or garbage is written like here:

```
SpawnsCrate0.Type=randomtext
SpawnsCrate0.Weight=1
```

I would like to also suggest to default to moneycrate not working powerups and unlist them from documentation.
Tested with [ra2yrbf](https://github.com/mbnq/RA2YRBF/releases/tag/v0.8.16)

I will test it in mp soon.

**Comment by Starkku** [comment]
CrateType not supported is treated as if nothing was parsed.

`darkness` not working was a parsing bug I fixed.

Many Powerups not working is an existing problem that I have not addressed with these changes.

File: src/Utilities/TemplateDef.h [hunk 1]
```diff
@@ -594,6 +594,99 @@ namespace detail
...
+				value = Powerup::Gas;
+			}
+			else if (!_strcmpi(parser.value(), "tiberium"))
+			{
+				value = Powerup::Tiberium;
+			}
+			else if (!_strcmpi(parser.value(), "pod"))
+			{
+				value = Powerup::Pod;
+			}
```
Comment (chaserli) [inline src/Utilities/TemplateDef.h:677]: you don't really need to do all these yourself, there is a string array in the game indexed by the enum value, if you look into RulesClass::Read_Powerup you'll see

File: src/Ext/WarheadType/Body.cpp [hunk 2]
```diff
@@ -243,6 +243,44 @@ void WarheadTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
...
 	);
+
+	char tempBuffer[32];
+
+	for (size_t i = 0; ; i++)
+	{
+		Nullable<Powerup> crateType;
+		_snprintf_s(tempBuffer, sizeof(tempBuffer), "SpawnsCrate%u.Type", i);
+		crateType.Read(exINI, pSection, tempBuffer);
+
```
Comment (MortonPL) [inline src/Ext/WarheadType/Body.cpp:254]: This should be outside of the loop.

File: src/Ext/WarheadType/Body.cpp [hunk 3]
```diff
@@ -243,6 +243,44 @@ void WarheadTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
...
+		if (i == 0 && !crateType.isset())
+			crateType.Read(exINI, pSection, "SpawnsCrate.Type");
+
+		if (!crateType.isset())
+			break;
+
+		if (crateType.isset())
```
Comment (MortonPL) [inline src/Ext/WarheadType/Body.cpp:261]:
```suggestion
		if (!crateType.isset())
			break;
```
No need to check twice.

File: src/Ext/WarheadType/Body.cpp [hunk 4]
```diff
@@ -243,6 +243,44 @@ void WarheadTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
...
+				this->SpawnsCrate_Types[i] = crateType;
+			else
+				this->SpawnsCrate_Types.push_back(crateType);
+
+			Nullable<int> weight;
+			_snprintf_s(tempBuffer, sizeof(tempBuffer), "SpawnsCrate%u.Weight", i);
+			weight.Read(exINI, pSection, tempBuffer);
+
+			if (i == 0 && !weight.isset())
+				weight.Read(exINI, pSection, "SpawnsCrate.Weight");
```
Comment (MortonPL) [inline src/Ext/WarheadType/Body.cpp:273]: This should be outside of the loop, too.

File: src/Utilities/TemplateDef.h (hunk 1 above)
Comment (MortonPL) [reply src/Utilities/TemplateDef.h:677]: Specifically, `char* CrateNames[]` at `0x7E523C`.

**Review by MortonPL — APPROVED** [review]
This looks good besides a few tiny inefficiencies in parsing.

**Pull request merged**